pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -231,11 +232,16 @@ func (rc *reconciler) reconcile() {
 					metrics.RecordForcedDetachMetric()
 					klog.Warningf(attachedVolume.GenerateMsgDetailed("attacherDetacher.DetachVolume started", fmt.Sprintf("This volume is not safe to detach, but maxWaitForUnmountDuration %v expired, force detaching", rc.maxWaitForUnmountDuration)))
 				}
-			}
-			if err != nil && !exponentialbackoff.IsExponentialBackoff(err) {
-				// Ignore exponentialbackoff.IsExponentialBackoff errors, they are expected.
-				// Log all other errors.
-				klog.Errorf(attachedVolume.GenerateErrorDetailed("attacherDetacher.DetachVolume failed to start", err).Error())
+			} else {
+				if !nestedpendingoperations.IsAlreadyExists(err) {
+					// add volume back to ReportAsAttached list
+					rc.actualStateOfWorld.AddVolumeToReportAsAttached(attachedVolume.VolumeName, attachedVolume.NodeName)
```

Comment:
If I understand correctly, this only handles the case where previous operation was already *in-progress*. But that may or may be the case and regardless there are checks in place to prevent concurrent operation on same volume. So IMO - this does not fixes the bug. Did I miss anything?

Reply:
This will handle the case where previous operation was  **not** in-progress. If the error is **in-progress error**,we will ignore it (**in-progress error** means this volume is maybe detaching, other errors mean this volume detach failed). If other error occurs, will add back this volume into attached list

Reply:
ah cool. I missed that. Yeah - I think either your or @jingxu97 's solution works. One advantage with this PR is - it is smaller.

Reply:
Yep

Reply:
minor nit : This requires little bit more explanation in comments about why we are adding volume back to actual state of the world. Currently the comment only says what is already evident from the code. (It almost just says call `AddVolumeToReportAsAttached`).

We need to say :

> Add the volume back to actual state of the world, because we removed it from actual state of world before calling detach and since we could not start Detach operation and there were no in-progress detach operations, the volume *must* be restored to its previous state in actual state of the world. 

One potential issue though here is - will this cause a volume which was previously marked "uncertain" (and hence not reported into node's status) as "certain" ? I think we should at very minimum honour the previously stored attachment state.